bpo-42207: Make llvm profile data location reference absolute

serge-sans-paille opened the pull request:
Otherwise, when running the testsuite, test_peg_generator tries to compile C
code using the optimized flags and fails because it cannot find the profile
data.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-42207](https://bugs.python.org/issue42207) -->
https://bugs.python.org/issue42207
<!-- /issue-number -->

vstinner commented on configure.ac:
```diff
@@ -1460,8 +1460,8 @@ case $CC in
     PGO_PROF_GEN_FLAG="-fprofile-instr-generate"
-    PGO_PROF_USE_FLAG="-fprofile-instr-use=code.profclangd"
-    LLVM_PROF_MERGER="${LLVM_PROFDATA} merge -output=code.profclangd *.profclangr"
+    PGO_PROF_USE_FLAG="-fprofile-instr-use=$PWD/code.profclangd"
```
I would prefer to use $(abs_builddir): I am not sure that $PWD is portable on any shell.

Would you mind to add a comment referecing bpo-42207 to explain than an absolute path allows to run the test suite outside the build directory?

Is the problem that pegen uses "NODIST" variables like CFLAGS_NODIST and LDFLAGS_NODIST?

cc @stratakis @pablogsal @lysnikolaou

pablogsal replied:
I would even prefer to deactivate this test on pgo builds or if detects that is running under a profile generation run given that it adds nothing useful and it takes a considerable time to execute.

vstinner replied:
test_peg_generator is not part of the default profile task (`./python -m test --pgo`: test_peg_generator is not part of Lib/test/libregrtest/pgo.py). The problem is once Python it built, running test_peg_generator outside the build directory fails, according to https://bugs.python.org/issue42207

pablogsal replied:
I know, what I am proposing is to skip it if it **ever** runs under a profile generation step

serge-sans-paille replied:
Two possible approaches: https://sergesanspaille.fedorapeople.org/0001-Disable-peg-generator-test-under-PGO.patch and https://sergesanspaille.fedorapeople.org/0001-Fix-peg-parser-in-presence-of-profile-data.patch , which one do you prefer?

pablogsal replied:
I assume that the first approach will be without the "Monkey patch CFLAGS" part, no?

In that case, I think I prefer the first one, as this test is particularly heavy with absolutely no gain on profile-based builds, it will benefit folks running the full test suite for PGO

serge-sans-paille replied:
@pablogsal your assumption was correct. I picked that patch and updated the PR accordingly.

serge-sans-paille committed:
Disable peg generator tests when building with PGO

Otherwise, when running the testsuite, test_peg_generator tries to compile C
code using the optimized flags and fails because it cannot find the profile
data.

https://bugs.python.org/issue42207
